### PR TITLE
Add GPU support for DNNL endpoint

### PR DESCRIPTION
**Pull request by georgen117**
**Description**: Describe your changes.

This enables the DNNL endpoint to use Intel GPU when training a model with Relu, Convolution, or MaxPool operators.

**Motivation and Context**
- Why is this change required? What problem does it solve?

A recent commit introduced ReluGrad, ConvGrad, and MaxPoolGrad ops for ORT training when using the DNNL endpoint. This makes it possible to use those ops with an Intel GPU.

### CODE REVIEW

**Comment by jywu-msft** [comment]
@mrry can you help take a look since you reviewed the previous PR's?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (mrry) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:315]: Checking my understanding here: is it intentional that these will be created for the CPU engine even when a GPU is available?

(I would have expected e.g. `diff_dst_mem_` to contain a reference to GPU memory when a GPU is available, and the same member to point to CPU memory otherwise. This seems to be how `primitive_dst_mem_` is handled. But I'm not familiar with how this is actually implemented....)

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (mrry) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:11]: Should this be "MaxPoolGrad"?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (mrry) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:318]: Style nit: for all of these `make_unique` calls, I think you shouldn't need to write the type name type. e.g. The following should work:

```c++
src_mem_ = onnxruntime::make_unique<dnnl::memory>(conv_bwd_weights_pd_.get()->src_desc(), cpu_engine, nullptr);
```

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (mrry) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:162]: Style nit: The following would be a little more compact... (I see this pattern in a few places).

```c++
dnnl::memory::desc pd(source_desc_);
```

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (mrry) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:401]: Not in this PR, but I should have caught this last time... the ternary operator here is redundant.

```c++
bool is_2D = src_dims_mkl.size() == 4;
bool is_1D = src_dims_mkl.size() == 3;
```

**Review by mrry — CHANGES_REQUESTED** [review]
Thanks George... mostly just a few nits on my end and a few things I'd like to clarify.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:70]: changing this to auto changes the iterator from a const_iterator to an iterator. The iterator is expected to be a const_iterator in the below code and using auto will cause a compilation error.

If OnnxRuntime suports C++17 then I could use `auto iter = std::as_const(dnnl_engine).find(dnnl::engine::kind::cpu);` I was not sure if OnnxRuntime supports C++17; so I went with the `const_iterator`.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:315]: This is describing the memory block that will be bound to in the Bind function. This is memory is from OnnxRuntime and exists in the CPU ram, not the GPU ram.  This memory will be copied to the GPU using the reorder below (Lines 327-328). So the cpu_engine is done purposely and is required.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:401]: I can update this.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:11]: fixed

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:162]: Thanks, fixed it in two places it showed up in this file.  I will have to keep my eye out for this pattern when doing code cleanup.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:318]: @mrry This is one of those things that stuck around because I was following the style of past developers.  I looked at the code in other files, for this endpoint and this style nit is everywhere. I am already working on a cleanup commit that only does code cleanup.  This would be better addressed in that commit instead of this commit.

File: cmake/external/dnnl.cmake
Comment (georgen117) [reply cmake/external/dnnl.cmake:4]: Thanks for asking about this.  I double-checked the code and we are good with this change.

**Comment by georgen117** [comment]
@mrry is there anything I have missed? I feel like I have addressed all of your comments.  

There were a few of the nits pick changes that I chose not to make in this commit but want to make in a follow-up commit.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (mrry) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:70]: Thanks for explaining!

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (mrry) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:318]: Sounds great, thanks.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h
Comment (mrry) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:315]: Thanks for clarifying!

**Review by mrry — APPROVED** [review]

**Comment by mrry** [comment]
Thanks George, this should be good to merge once all the CI pipelines pass.

**Comment by georgen117** [comment]
> Thanks George, this should be good to merge once all the CI pipelines pass.

Sounds great.  I have a cleanup commit that I was waiting on this commit before doing a pull commit.  There are a few of your nit-picks that will be incorporated in that upcoming change.

**Comment by georgen117** [comment]
@mrry it looks like all of the checks have passed.  

Its looks like it's waiting on someone with write access to merge the pull request.

**Comment by mrry** [comment]
Thanks for the reminder :). Merging now!

**Comment by georgen117** [comment]
> Thanks for the reminder :). Merging now!

Your welcome.  I figured you were just waiting for the CI builds to complete. Thanks again for the review work it really is appreciated.

**Pull request merged**
